### PR TITLE
Allow routes from routes prop

### PR DESCRIPTION
**Pull request by alexanderchr**
As discussed in #252.

### CODE REVIEW

**Comment by alexanderchr** [comment]
Might actually have to do this for all child routes as well, will test.

File: src/patch.dev.js [hunk 1]
```diff
@@ -77,26 +77,43 @@ if (React.createElement.isPatchedByReactHotLoader) {
...
-    props &&
-    typeof props.component === 'function'
-  ) {
-    // Side effect 😱
-    // Force proxies to update since React Router ignores new props.
-    resolveType(props.component);
+  if (isReactRouterish(type) && props) {
+    // A router's routes may be given as children or through
+    // the `routes` prop.
+    var topRoutes = [].concat(props.routes, props.children).filter(Boolean);
```
Comment (gaearon) [inline src/patch.dev.js]: Let’s move this initial loop into `forceUpdateRouteAndChildRouteComponents` as well.

File: src/patch.dev.js [hunk 2]
```diff
@@ -77,26 +77,43 @@ if (React.createElement.isPatchedByReactHotLoader) {
...
+    // Side effect 😱
+    // Force proxies to update since React Router ignores new props.
+    resolveType(route.component);
+  }
+
+  // Child routes will be in `props.children` when defining routes using react
+  // elements and in `childRoutes` when using plain routes
+  var childRoutes = [].concat(route.childRoutes, route.props && route.props.children).filter(Boolean);
+
+  for (var _len = childRoutes.length, _key = 0; _key < _len; _key++) {
```
Comment (gaearon) [inline src/patch.dev.js]: Let’s just use `for (var i = 0; i < childRoutes.length; i++)`?

File: src/patch.dev.js (hunk 1 above)
Comment (alexanderchr) [inline src/patch.dev.js]: Sure.

File: src/patch.dev.js [hunk 3]
```diff
@@ -77,26 +77,42 @@ if (React.createElement.isPatchedByReactHotLoader) {
...
+  // elements and in `routes` or `childRoutes` when using plain routes
+  var childRoutes = []
+    .concat(
+      route.routes,
+      route.childRoutes,
+      route.props && route.props.children)
+    .filter(Boolean);
+
+  for (var i = 0; i < childRoutes.length; i++) {
+    var childRoute = childRoutes[_key];
```
Comment (gaearon) [inline src/patch.dev.js]: Still using `_key` here

Comment (alexanderchr) [inline src/patch.dev.js]: Yeah, nice catch. It seems like it worked because babel were transpiling it to _key anyways.

File: src/patch.dev.js [hunk 4]
```diff
@@ -77,26 +77,42 @@ if (React.createElement.isPatchedByReactHotLoader) {
...
+  // TODO: check whether it is possible to also handle the `getComponent` case here
+  if (route.component && typeof route.component === 'function') {
+    // Side effect 😱
+    // Force proxies to update since React Router ignores new props.
+    resolveType(route.component);
+  }
+
+  // Child routes will be in `props.children` when defining routes using react
+  // elements and in `routes` or `childRoutes` when using plain routes
+  var childRoutes = []
```
Comment (gaearon) [inline src/patch.dev.js]: Minor nit: let’s use `const` where possible, `let` otherwise.

Comment (alexanderchr) [inline src/patch.dev.js]: Sure.

File: src/patch.dev.js [hunk 5]
```diff
@@ -77,26 +77,43 @@ if (React.createElement.isPatchedByReactHotLoader) {
...
-    type.name === 'IndexRoute'
+    type.displayName === 'Router' ||
+    type.name === 'Router' // In case Ryan and Michael embrace ES6 classes
   );
 }
 
+function forceUpdateComponentOfRouteAndChildRoutes(route) {
+  // TODO: check whether it is possible to also handle the `getComponent` case here
+  const components = []
+    .concat(route.component, route.components && Object.values(route.components))
```
Comment (gaearon) [inline src/patch.dev.js]: `Object.values` may not work in older browsers, we shoot for wide compatibility.

File: src/patch.dev.js [hunk 6]
```diff
@@ -77,26 +77,43 @@ if (React.createElement.isPatchedByReactHotLoader) {
...
+    .filter(c => typeof(c) === 'function');
+
+  for(const component of components) {
```
Comment (gaearon) [inline src/patch.dev.js]: Nit: space after `for`.

File: src/patch.dev.js (hunk 5 above)
Comment (alexanderchr) [inline src/patch.dev.js]: Right, changed it a bit.

**Comment by gaearon** [comment]
Thanks, this looks good. I have a few more style nits but I’m happy to fix those myself. I need to add a linter anyway.

Would you be open to considering adding some tests for this? I have a basic Mocha setup for the Babel plugin, and it seems like this is another example of something we want to be tested.

File: src/hackRouter.js [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+function objectValues(obj) {
+  let values = [];
+  for (const key in obj) {
+    if (obj.hasOwnProperty(key)) {
+      values.push(obj[key]);
+    }
+  }
+  return values;
+}
+
```
Comment (alexanderchr) [inline src/hackRouter.js]: Might want to use a real polyfill instead

**Comment by alexanderchr** [comment]
Sure, added a few tests.

Also had some midnight inspiration and rewrote all of it. I included these changes in the last commit, so feel free to discard them if you think the old way was better. Otherwise you can just squash them into the first refactor commit. Tests should hopefully be passing either way.

File: src/hackRouter.js (hunk 7 above)
Comment (gaearon) [inline src/hackRouter.js]: Maybe we’d just use `Object.keys(obj).map(key => obj[key])`?

**Comment by gaearon** [comment]
This looks great! `hackRouter` sounds a little intimidating, let’s rename it to `addReactRouterFix()` or something similar?

**Comment by alexanderchr** [comment]
Sure!

File: src/hackRouter.js (hunk 7 above)
Comment (alexanderchr) [inline src/hackRouter.js]: Yeah, that's it. Thanks.

File: src/fixupReactRouter.js [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    .filter(c => typeof(c) === 'function');
+}
+
+function fixupReactRouter(props, forceUpdateFunction) {
+  // Ideally we want to teach React Router to receive children.
+  // We're not in a perfect world, and a dirty workaround works for now.
+  // https://github.com/reactjs/react-router/issues/2182
+  const routes = props.routes || ensureArray(props.children);
+  const components = componentsToForceUpdate(routes);
+  for (const component of components) {
```
Comment (gaearon) [inline src/fixupReactRouter.js]: [`for of` requires `Symbol`s](http://babeljs.io/repl/#?evaluate=true&presets=es2015%2Creact&experimental=false&loose=false&spec=false&code=for%20(var%20x%20of%20y)%20%7B%0A%20%20%0A%7D), let’s use `components.forEach`

File: src/fixupReactRouter.js [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+function isReactRouterish(type) {
+  return type && (
+    type.displayName === 'Router' ||
+    type.name === 'Router' // In case Ryan and Michael embrace ES6 classes
+  );
+}
+
+function componentsToForceUpdate(routes) {
```
Comment (gaearon) [inline src/fixupReactRouter.js]: Let’s rename to `extractComponents`. We’re not exactly force updating them anyway.

**Comment by gaearon** [comment]
Out in `3.0.0-alpha.13`, thank you!

**Comment by alexanderchr** [comment]
Cool. Thanks for reviewing!

**Pull request merged**